### ISSUE TITLE
Finetune ClipSeg model 

### ISSUE THREAD

**Issue by sleeping4cat** [huggingface/transformers#24494]
### Feature request

Quite recently, I was exploring zero-shot classification to segment medical images. And it looks quite promising. I stumbled upon ```ClipSeg``` a few days ago and it looked wonderful and just well-suited for my work. Unfortunately, I couldn't find any tutorials or notebooks that showed how to perform fine-tuning on ClipSeg model. 

I am assuming, we have to train the decoder with a dataset containing binary classification images of cells and their corresponding masks and a text description. Unfortunately, a bit confused. is there any tutorials/resources anyone could suggest on this topic? Cuz I couldn't none. 

### Motivation

```ClipSeg``` shows a lot of potential than SAM (Segment Anything Model). Unfortunately, there's no fine-tuning script neither instructions on **How to prepare the dataset?** which is very frustrating. Will love some help from the community. 

And another point, Zero shot classification looks a way lot better option with fine-tuning than training a model like ```U-Net```, ```R-CNN``` and others from scratch while you have very few images and don't have much room to play around with.

### Your contribution

I could provide a PR on my LinkedIn, where I have a lot of AI experts as my connections and then I contribute in the programming as well.

**Comment by sgugger** [comment]
cc @alaradirik for information.

**Comment by sleeping4cat** [comment]
@sgugger Can you provide some insights/help/update to my requested feature?

**Comment by sgugger** [comment]
cc @amyeroberts and @rafaelpadilla

**Comment by qgallouedec** [comment]
Hey @sleeping4cat, sorry for the very slow reply, we're catching up on the backlog.

Nothing is missing on the modelling side: `CLIPSegForImageSegmentation` already returns a loss when you pass `labels`, so fine-tuning the decoder is a plain Trainer / training loop.

```python
import numpy as np
import torch
from PIL import Image, ImageDraw
from torch.utils.data import Dataset
from transformers import CLIPSegForImageSegmentation, CLIPSegProcessor, Trainer, TrainingArguments


model_id = "CIDAS/clipseg-rd64-refined"
processor = CLIPSegProcessor.from_pretrained(model_id)
model = CLIPSegForImageSegmentation.from_pretrained(model_id)

# Train the decoder only, the CLIP backbone stays frozen
model.clip.requires_grad_(False)


class SegmentationDataset(Dataset):
    """Each item is an image, a text prompt, and a binary mask of the same size as the image."""

    def __init__(self, size=16):
        self.size = size

    def __len__(self):
        return self.size

    def __getitem__(self, idx):
        image = Image.new("RGB", (352, 352), "black")
        mask = Image.new("L", (352, 352), 0)
        box = [80 + idx, 60 + idx, 250 + idx, 240 + idx]
        ImageDraw.Draw(image).ellipse(box, fill="white")
        ImageDraw.Draw(mask).ellipse(box, fill=255)
        return {"image": image, "text": "a circle", "mask": mask}


def collate(examples):
    # The decoder predicts logits of shape (batch_size, 352, 352), so masks are resized to 352x352
    # and cast to float in [0, 1]: the loss is a BCEWithLogitsLoss over the mask.
    inputs = processor(
        text=[example["text"] for example in examples],
        images=[example["image"] for example in examples],
        padding=True,
        return_tensors="pt",
    )
    masks = [np.array(example["mask"].resize((352, 352), Image.NEAREST)) for example in examples]
    inputs["labels"] = torch.from_numpy(np.stack(masks)).float() / 255.0
    return inputs


trainer = Trainer(
    model=model,
    args=TrainingArguments(output_dir="clipseg-finetuned", num_train_epochs=2, learning_rate=1e-4, remove_unused_columns=False),
    train_dataset=SegmentationDataset(),
    data_collator=collate,
)
trainer.train()

# Inference with the fine-tuned decoder
image = SegmentationDataset()[0]["image"]
inputs = processor(text=["a circle"], images=[image], return_tensors="pt").to(model.device)
with torch.no_grad():
    logits = model(**inputs).logits
predicted_mask = logits.sigmoid() > 0.5
```

For the dataset, `labels` are float binary masks of shape `(batch_size, height, width)` matching the logits (same spatial size as the `pixel_values` you feed in, 352x352 with the default processor), not class indices. The docstring currently says `LongTensor` of shape `(batch_size,)`, which is wrong and will raise; we'll fix that.
